File: 0069-Sqrt(x)/0069.py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x <= 1: return x
        left, right = 0, x
        res = 0

        while left <= right:
            mid = (left + right) // 2
            # or mid <= x / mid to to avoid potential overflow issues when x is large
            if mid * mid <= x:
                res = mid
                left = mid + 1
            else:
                right = mid - 1
        
        return res

File: 0069-Sqrt(x)/0069.java
class Solution {
    public int mySqrt(int x) {
        if (x < 2)
            return x;
        int left = 0, right = x, res = 0;

        while (left <= right) {
            int mid = (left + right) / 2;

            if (mid <= x / mid) {
                res = mid;
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return res;
    }
}